### PR TITLE
[FEAT] Add support for pushing the amplification results in a MongoDB database

### PR DESCRIPTION
**Pull request by Tailp**
Hi , related issue (#843) . Ready for merging after done reviewing and change requests :).

### CODE REVIEW

**Comment by danglotb** [comment]
The changes look good.

Could create a dedicated method for updating the database? _i.e._ not doing all the database stuff in `reportStd`, since this method should only output on the standard output.

**Comment by Tailp** [comment]
Hi @danglotb :), you mean an update method which only update an already existed document document ? . If that's the case then they already have this updateOne or updateMany method which take in a filter query and a new document . More details [here](https://mongodb.github.io/mongo-java-driver/3.10/driver/getting-started/quick-start/). Basicly it would look like this
```
mongoClient = connectToMongo(this.mongoUrl);
MongoDatabase database = getDatabase(this.dbName,mongoClient);
MongoCollection<Document> coll = getCollection(this.colName,database);
coll.updateOne(FILTER,new Document("$set", NEWDOC));
```
Sorry if I missunderstood something , please point out if that's the case :) .

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java [hunk 1]
```diff
@@ -235,6 +239,14 @@ public TestSelectorElementReport report() {
     }
 
     private String reportStdout() {
+        if (MongodbManager.getInstance().getDbConnectable()) {
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: Hi @Tailp,

I would like the following if block in a dedicated method, like `reportMongoDB()`.

Basically, create dedicated methods for all actions that are done on the mongoDB, in order to separate the mongoDB management from the rest of the code.

Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: ah I think I understand your point :). I made some changes just now. Hopefully this passes the test on the CI  and is what you wanted.

**Comment by monperrus** [comment]
Per our discussion today, we will split this PR in two

* the first one to introduce a DspotInformationCollector infrastructure
* the second one to introduce the optional dependency to MongoDB

File: dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java [hunk 2]
```diff
@@ -222,6 +223,10 @@ public TestSelectorElementReport report() {
                     .append("%")
                     .append(AmplificationHelper.LINE_SEPARATOR);
             lastReport = new TestSelectorElementReportImpl(report.toString(), jsonReport(coverageResults));
+
+            // Sending data to Mongodb
+            MongodbManager.getInstance().reportJacocoMongoDb(this.currentClassTestToBeAmplified.getQualifiedName(),"" + this.initialCoverage.getInstructionsCovered(),"" + coverageResults.getInstructionsCovered(),"" + coverageResults.getInstructionsTotal());
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java]: collector.reportInitInformation()

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java [hunk 3]
```diff
@@ -235,6 +236,8 @@ public TestSelectorElementReport report() {
     }
 
     private String reportStdout() {
+        MongodbManager.getInstance().reportPitMutantMongoDB(this.currentClassTestToBeAmplified.getQualifiedName(),"" + this.originalKilledMutants.size(),"" + this.getNbTotalNewMutantKilled());
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: collector.reportMutantInformation()

File: dspot/src/main/java/eu/stamp_project/Main.java [hunk 4]
```diff
@@ -59,6 +60,10 @@ public static void run() {
 		// global report handling
 		Main.GLOBAL_REPORT.output();
 		Main.GLOBAL_REPORT.reset();
+		// Send info to Mongodb
+		if (MongodbManager.getInstance().getDbConnectable()) {
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/Main.java]: collector.reportInitInformation()

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java (hunk 1 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: hi @danglotb, I put all codeblocks into reportMongoDB() functions :) . I would be nice if you have some spare time to check if it's ok to go or if some more fixes I need to do. Thank you for your patience.

**Comment by Tailp** [comment]
Will open a new PR for this :) .

**Pull request closed without merging**